File: packages/types/src/patterns.ts
export type PatternDetectionType =
  | 'CIRCULAR_FUND_FLOW'
  | 'BURNER_SIM'
  | 'NETWORK_HUB'
  | 'BRIDGE_ENTITY'
  | 'RAPID_RELATIONSHIP_EXPANSION';

export type PatternDetectionSeverity = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface PatternDetectionResult {
  id: string;
  investigation_id: string;
  pattern_type: PatternDetectionType;
  severity: PatternDetectionSeverity;
  confidence: number;
  title: string;
  description: string;
  entity_ids: string[];
  relationship_ids: string[];
  evidence_ids: string[];
  event_ids: string[];
  metadata: Record<string, unknown>;
  detected_at: string;
}

export interface PatternDetectionResponse {
  investigation_id: string;
  patterns: PatternDetectionResult[];
}

// ============================================================
// PHASE C — UNIFIED PATTERNS WORKSPACE CONTRACT
// ============================================================
// The Patterns workspace presents the backend detection engine output
// (`PatternDetectionResult`) enriched with investigation-scoped context
// (`entityRefs` / `evidenceRefs` / `relationshipRefs`). Every field traces to
// a real backend value or the investigation's persisted rows. Fields the
// backend cannot supply are simply absent — never invented.
// ============================================================

/** Entity context referenced by a pattern, resolved from the
 *  investigation-scoped persisted entity rows. */
export interface PatternEntityRef {
  id: string;
  name: string;
  type: string;
}

/** Evidence context referenced by a pattern, resolved from the
 *  investigation-scoped persisted evidence rows. */
export interface PatternEvidenceRef {
  id: string;
  title: string;
}

/** Relationship context referenced by a pattern, resolved from the
 *  investigation-scoped persisted relationship rows. */
export interface PatternRelationshipRef {
  id: string;
  sourceName?: string;
  targetName?: string;
  type?: string;
}

/**
 * Unified investigation-scoped pattern artifact rendered by the Patterns
 * workspace. Extends the wire contract with context enrichment.
 *
 * Sourcing rules:
 *  - `typeLabel`        : API mode → the authoritative backend `pattern_type`
 *                         term (never renamed); mock mode → the fixture label.
 *  - `status`           : mock fixture only; absent in API mode (the relational
 *                         model has no review queue, so a status is honest only
 *                         when the source carries one).
 *  - `metrics`          : API mode → deterministic rows derived from the backend
 *                         `metadata` (real analytic values); mock mode → fixture
 *                         metrics.
 *  - `timeAgo`          : API mode → computed from the real `detected_at`;
 *                         mock mode → fixture value.
 *  - `entityRefs` / `evidenceRefs` / `relationshipRefs` : resolved from the
 *    investigation's persisted rows; empty when unavailable.
 */
export interface PatternArtifact extends PatternDetectionResult {
  typeLabel: string;
  status?: 'new' | 'reviewing' | 'dismissed' | 'confirmed';
  metrics: Record<string, string>;
  timeAgo: string;
  entityRefs: PatternEntityRef[];
  evidenceRefs: PatternEvidenceRef[];
  relationshipRefs: PatternRelationshipRef[];
}

File: apps/api/app/schemas/real/patterns.py
"""Typed contracts for investigation-scoped suspicious pattern detection."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import Field

from app.schemas.common import SchemaBase


class PatternType(StrEnum):
    CIRCULAR_FUND_FLOW = "CIRCULAR_FUND_FLOW"
    BURNER_SIM = "BURNER_SIM"
    NETWORK_HUB = "NETWORK_HUB"
    BRIDGE_ENTITY = "BRIDGE_ENTITY"
    RAPID_RELATIONSHIP_EXPANSION = "RAPID_RELATIONSHIP_EXPANSION"


class PatternSeverity(StrEnum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class PatternDetectionResult(SchemaBase):
    id: str
    investigation_id: UUID
    pattern_type: PatternType
    severity: PatternSeverity
    confidence: float = Field(ge=0.0, le=1.0)
    title: str
    description: str
    entity_ids: list[UUID] = Field(default_factory=list)
    relationship_ids: list[UUID] = Field(default_factory=list)
    evidence_ids: list[UUID] = Field(default_factory=list)
    event_ids: list[UUID] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    detected_at: datetime


class PatternDetectionResponse(SchemaBase):
    investigation_id: UUID
    patterns: list[PatternDetectionResult] = Field(default_factory=list)
